File: 0882. Reachable Nodes In Subdivided Graph/Solution.py
from collections import defaultdict
import heapq
from typing import List


class Solution:
    def reachableNodes(self, edges: List[List[int]], maxMoves: int, n: int) -> int:
        E = defaultdict(dict)
        for u, v, cnt in edges:
            E[u][v] = E[v][u] = cnt + 1
        seen = {}
        pq = [(-maxMoves, 0)]
        while pq:
            remain_moves, node = heapq.heappop(pq)
            if node not in seen:
                seen[node] = remain_moves = -remain_moves
                for neighbour, cnt in E[node].items():
                    if neighbour not in seen and remain_moves - cnt >= 0:
                        heapq.heappush(pq, (cnt - remain_moves, neighbour))
        res = len(seen)
        for u, v, cnt in edges:
            res += min(seen.get(u, 0) + seen.get(v, 0), cnt)
        return res

File: 0882. Reachable Nodes In Subdivided Graph/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public int reachableNodes(int[][] edges, int maxMoves, int n) {
        Map<Integer, Map<Integer, Integer>> E = new HashMap<>();
        for (int i = 0; i < n; i++) {
            E.put(i, new HashMap<>());
        }
        for (int[] edge : edges) {
            E.get(edge[0]).put(edge[1], edge[2] + 1);
            E.get(edge[1]).put(edge[0], edge[2] + 1);
        }

        Map<Integer, Integer> seen = new HashMap<>();
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.offer(new int[]{maxMoves, 0});
        while (!pq.isEmpty()) {
            int remainMoves = pq.element()[0], node = pq.element()[1];
            pq.remove();
            if (!seen.containsKey(node)) {
                seen.put(node, remainMoves);
                for (Integer neighbour : E.get(node).keySet()) {
                    if (!seen.containsKey(neighbour) && E.get(node).get(neighbour) <= remainMoves) {
                        pq.offer(new int[]{remainMoves - E.get(node).get(neighbour), neighbour});
                    }
                }
            }
        }
        int res = seen.size();
        for (int[] edge : edges) {
            int a = seen.getOrDefault(edge[0], 0);
            int b = seen.getOrDefault(edge[1], 0);
            res += Math.min(a + b, edge[2]);
        }
        return res;
    }
}
